Deprecate callback setting variant price from master

waiting-for-dev opened the pull request:
This simplifies logic, and it can be seen as preparatory work to propose
to the user the inheritance of all, and not just the default, prices
from master when creating a new variant.

Extracted from #3994 

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [ ] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

kennyadsl reviewed (approved):
Thanks Marc!

kennyadsl commented:
Maybe we can add a spec that tests the deprecation warning emitted when the price is not passed?

spaghetticode commented on sample/db/samples/variants.rb:
```diff
@@ -30,184 +30,218 @@
     sku: "SOL-00003",
-    cost_price: 17
+    cost_price: 17,
+    price: 10
```
it seems a bit strange that `price` is lower than the `cost_price`, shouldn't it be greater than that?

waiting-for-dev replied:
Fair enough, fixed :sweat_smile:

waiting-for-dev commented:
> Maybe we can add a spec that tests the deprecation warning emitted when the price is not passed?

Makes sense. I added it and rebased from master.
